dbms/src/Server/Server.cpp

```diff
@@ -941,6 +880,11 @@ int Server::main(const std::vector<std::string> & /*args*/)
     {
         LOG_INFO(log, "Using format_version={} (default settings).", STORAGE_FORMAT_CURRENT.identifier);
     }
+    if (storage_config.s3_config.isS3Enabled() && storage_config.format_version != STORAGE_FORMAT_V5.identifier)
+    {
+        LOG_WARNING(log, "'storage.format_version' must be set to 5 when S3 is enabled!");
+        throw Exception("'storage.format_version' must be set to 5 when S3 is enabled!");
+    }
```

Comment:
This would be better if we advance the format_version to 6 later, we don't need to let the user remove the storage.format_version configuration item from the configuration file? @flowbehappy 
```
    if (storage_config.format_version)
    {
        if (storage_config.s3_config.isS3Enabled() && storage_config.format_version != STORAGE_FORMAT_V5.identifier)
        {
            LOG_WARNING(log, "'storage.format_version' must be set to 5 when S3 is enabled!");
            throw Exception("'storage.format_version' must be set to 5 when S3 is enabled!");
        }
        setStorageFormat(storage_config.format_version);
        LOG_INFO(log, "Using format_version={} (explicit storage format detected).", storage_config.format_version);
    }
    else
    {
        if (storage_config.s3_config.isS3Enabled())
        {
            // If the user does not explicitly set format_version in the config file but
            // enables S3, then we set up a proper format version to support S3.
            setStorageFormat(5);
            LOG_INFO(log, "Using format_version={} (inferr by S3 is enabled).", STORAGE_FORMAT_V5.identifier);
        }
        else
        {
            // Use the default settings
            LOG_INFO(log, "Using format_version={} (default settings).", STORAGE_FORMAT_CURRENT.identifier);
        }
    }
```

Reply:
LGTM

Reply:
changed